Index: gtk+2.0-2.14.0/Makefile.am
===================================================================
--- gtk+2.0-2.14.0.orig/Makefile.am	2008-09-04 06:36:14.000000000 +0200
+++ gtk+2.0-2.14.0/Makefile.am	2008-09-05 00:36:22.000000000 +0200
@@ -1,7 +1,12 @@
 ## Makefile.am for GTK+
 include $(top_srcdir)/Makefile.decl
 
-SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests perf contrib
+# don't build perf measurement tools in static builds
+if !ENABLE_STATIC
+MAYBE_PERF = perf
+endif
+
+SRC_SUBDIRS = gdk-pixbuf gdk gtk modules demos tests $(MAYBE_PERF) contrib
 SUBDIRS = po po-properties $(SRC_SUBDIRS) docs m4macros
 
 # require automake 1.4
