Index: gtk+2.0-2.13.3/gdk-pixbuf/io-ico.c
===================================================================
--- gtk+2.0-2.13.3.orig/gdk-pixbuf/io-ico.c	2008-06-18 14:58:49.000000000 +0200
+++ gtk+2.0-2.13.3/gdk-pixbuf/io-ico.c	2008-06-18 14:58:54.000000000 +0200
@@ -231,6 +231,14 @@
 	
  	State->HeaderSize = 6 + IconCount*16;
 
+	if (State->HeaderSize < 0) {
+		g_set_error (error,
+			     GDK_PIXBUF_ERROR,
+			     GDK_PIXBUF_ERROR_CORRUPT_IMAGE,
+			     _("Invalid header in icon"));
+		return;
+	}
+
  	if (State->HeaderSize>State->BytesInHeaderBuf) {
  		guchar *tmp=g_try_realloc(State->HeaderBuf,State->HeaderSize);
 		if (!tmp) {
