Debian #381081; GNOME #447118

Index: gtk+2.0-2.15.2/gdk/directfb/gdkdrawable-directfb.c
===================================================================
--- gtk+2.0-2.15.2.orig/gdk/directfb/gdkdrawable-directfb.c	2009-01-27 05:03:31.000000000 +0100
+++ gtk+2.0-2.15.2/gdk/directfb/gdkdrawable-directfb.c	2009-01-27 11:21:00.000000000 +0100
@@ -47,6 +47,7 @@
 #include "gdkregion-generic.h"
 #include "gdkalias.h"
 
+#include <cairo.h>
 #include "cairo-directfb.h"
 
 
