Debian #381081; GNOME #447118

Index: gtk+2.0-2.14.0/gdk/directfb/gdkdrawable-directfb.c
===================================================================
--- gtk+2.0-2.14.0.orig/gdk/directfb/gdkdrawable-directfb.c	2008-09-04 06:36:10.000000000 +0200
+++ gtk+2.0-2.14.0/gdk/directfb/gdkdrawable-directfb.c	2008-09-05 00:36:24.000000000 +0200
@@ -47,6 +47,7 @@
 #include "gdkregion-generic.h"
 #include "gdkalias.h"
 
+#include <cairo.h>
 #include "cairo-directfb.h"
 
 
