diff -Nur gtk+-2.8.16.orig/gtk/gtkentry.c gtk+-2.8.16/gtk/gtkentry.c
--- gtk+-2.8.16.orig/gtk/gtkentry.c	2006-03-08 15:19:31.000000000 +0100
+++ gtk+-2.8.16/gtk/gtkentry.c	2006-03-23 17:23:38.000000000 +0100
@@ -1000,7 +1000,7 @@
 
   entry->editable = TRUE;
   entry->visible = TRUE;
-  entry->invisible_char = '*';
+  entry->invisible_char = 0x25cf;
   entry->dnd_position = -1;
   entry->width_chars = -1;
   entry->is_cell_renderer = FALSE;
