diff -Nur gtk+-2.10.11.orig/gtk/gtksettings.c gtk+-2.10.11/gtk/gtksettings.c
--- gtk+-2.10.11.orig/gtk/gtksettings.c	2007-04-04 13:02:43.000000000 +0200
+++ gtk+-2.10.11/gtk/gtksettings.c	2007-04-04 13:03:11.000000000 +0200
@@ -268,7 +268,7 @@
                                              g_param_spec_string ("gtk-icon-theme-name",
 								  P_("Icon Theme Name"),
 								  P_("Name of icon theme to use"),
-								  "hicolor",
+								  "gnome",
 								  GTK_PARAM_READWRITE),
                                              NULL);
   g_assert (result == PROP_ICON_THEME_NAME);
