Index: gtk+2.0-2.14.0/configure.in
===================================================================
--- gtk+2.0-2.14.0.orig/configure.in	2008-09-05 00:36:26.000000000 +0200
+++ gtk+2.0-2.14.0/configure.in	2008-09-05 00:36:27.000000000 +0200
@@ -1758,7 +1758,7 @@
 CAIRO_PREFIX="`pkg-config --variable=prefix cairo`"
 
 if test $enable_explicit_deps != yes ; then
-  GDK_PIXBUF_EXTRA_LIBS=
+  GDK_PIXBUF_EXTRA_LIBS="$MATH_LIB"
 fi
 
 AC_SUBST(GDK_PIXBUF_PACKAGES)
