Index: gtk+2.0-2.15.2/configure.in
===================================================================
--- gtk+2.0-2.15.2.orig/configure.in	2009-01-27 11:21:03.000000000 +0100
+++ gtk+2.0-2.15.2/configure.in	2009-01-27 11:21:03.000000000 +0100
@@ -1751,7 +1751,7 @@
 CAIRO_PREFIX="`pkg-config --variable=prefix cairo`"
 
 if test $enable_explicit_deps != yes ; then
-  GDK_PIXBUF_EXTRA_LIBS=
+  GDK_PIXBUF_EXTRA_LIBS="$MATH_LIB"
 fi
 
 AC_SUBST(GDK_PIXBUF_PACKAGES)
