Index: gtk+2.0-2.14.0/gtk/gtksettings.c
===================================================================
--- gtk+2.0-2.14.0.orig/gtk/gtksettings.c	2008-09-04 06:35:27.000000000 +0200
+++ gtk+2.0-2.14.0/gtk/gtksettings.c	2008-09-05 00:36:27.000000000 +0200
@@ -301,7 +301,7 @@
                                              g_param_spec_string ("gtk-fallback-icon-theme",
 								  P_("Fallback Icon Theme Name"),
 								  P_("Name of a icon theme to fall back to"),
-								  NULL,
+								  "gnome",
 								  GTK_PARAM_READWRITE),
                                              NULL);
   g_assert (result == PROP_FALLBACK_ICON_THEME);
