Index: gtk+2.0-2.15.2/gtk/gtksettings.c
===================================================================
--- gtk+2.0-2.15.2.orig/gtk/gtksettings.c	2009-01-27 05:02:03.000000000 +0100
+++ gtk+2.0-2.15.2/gtk/gtksettings.c	2009-01-27 11:21:04.000000000 +0100
@@ -301,7 +301,7 @@
                                              g_param_spec_string ("gtk-fallback-icon-theme",
 								  P_("Fallback Icon Theme Name"),
 								  P_("Name of a icon theme to fall back to"),
-								  NULL,
+								  "gnome",
 								  GTK_PARAM_READWRITE),
                                              NULL);
   g_assert (result == PROP_FALLBACK_ICON_THEME);
