Index: gtk+2.0-2.12.9/modules/input/imviqr.c
===================================================================
--- gtk+2.0-2.12.9.orig/modules/input/imviqr.c	2008-03-12 05:17:03.000000000 +0100
+++ gtk+2.0-2.12.9/modules/input/imviqr.c	2008-03-29 20:34:14.000000000 +0100
@@ -244,7 +244,7 @@
   N_("Vietnamese (VIQR)"), /* Human readable name */
   GETTEXT_PACKAGE,	   /* Translation domain */
    GTK_LOCALEDIR,	   /* Dir for bindtextdomain (not strictly needed for "gtk+") */
-  "vi"			   /* Languages for which this module is the default */
+  ""			   /* Languages for which this module is the default */
 };
 
 static const GtkIMContextInfo *info_list[] = {
