Index: gtk+2.0-2.11.3/gdk-pixbuf/gdk-pixbuf-io.c
===================================================================
--- gtk+2.0-2.11.3.orig/gdk-pixbuf/gdk-pixbuf-io.c	2007-06-16 14:25:47.000000000 +0200
+++ gtk+2.0-2.11.3/gdk-pixbuf/gdk-pixbuf-io.c	2007-06-16 14:26:24.000000000 +0200
@@ -32,6 +32,7 @@
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
+#include <sys/utsname.h>
 
 #include "gdk-pixbuf-private.h"
 #include "gdk-pixbuf-io.h"
@@ -280,8 +280,20 @@
   gchar *result = g_strdup (g_getenv ("GDK_PIXBUF_MODULE_FILE"));
 
   if (!result)
-	  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gdk-pixbuf.loaders", NULL);
-
+    {
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+# if defined (__i386__)
+      result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gdk-pixbuf.loaders.32", NULL);
+# else
+      result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gdk-pixbuf.loaders.64", NULL);
+#endif
+      /* Prefer compat gdk-pixbuf.loaders file if it's usable. */
+      if (g_file_test(result, G_FILE_TEST_EXISTS))
+        return result;
+      g_free (result);
+#endif
+      result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gdk-pixbuf.loaders", NULL);
+    }
   return result;
 }
 
@@ -524,6 +525,21 @@
 
 #endif /* USE_GMODULE */
 
+static gchar*
+_gdk_pixbuf_libdir (void)
+{
+#if defined(__linux__) && defined (__i386__)
+	struct utsname uts;
+
+	uname(&uts);
+	if (!strcmp("x86_64", uts.machine)
+	    && !access("/usr/lib32/gtk-2.0", R_OK|X_OK))
+		return "/usr/lib32/gtk-2.0/2.10.0/loaders";
+	else
+#endif
+	return PIXBUF_LIBDIR;
+}
+
 static void 
 gdk_pixbuf_io_init (void)
 {
@@ -531,7 +547,9 @@
 	GIOChannel *channel;
 	gchar *gdkpixbuf_module_file_str = gdk_pixbuf_get_module_file ();
 	/* FIXME: using an absolute path requires patching Makefile.am */
-	gchar *gdkpixbuf_module_files_d_str = g_build_filename (PIXBUF_LIBDIR,
+	/*   ... and breaks using the 32bit lib on a 64 bit system \o/ */
+
+	gchar *gdkpixbuf_module_files_d_str = g_build_filename (_gdk_pixbuf_libdir(),
 							 "..",
 							 "loader-files.d",
 							 NULL);
Index: gtk+2.0-2.11.3/gtk/gtkrc.c
===================================================================
--- gtk+2.0-2.11.3.orig/gtk/gtkrc.c	2007-06-15 20:07:37.000000000 +0200
+++ gtk+2.0-2.11.3/gtk/gtkrc.c	2007-06-16 14:26:24.000000000 +0200
@@ -450,7 +450,20 @@
       if (im_module_file)
 	result = g_strdup (im_module_file);
       else
-	result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtk.immodules", NULL);
+        {
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+# if defined (__i386__)
+         result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtk.immodules.32", NULL);
+# else
+         result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtk.immodules.64", NULL);
+# endif
+          /* Prefer compat gtk.immodules file if it's usable. */
+          if (g_file_test(result, G_FILE_TEST_EXISTS))
+            return result;
+          g_free (result);
+#endif
+          result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtk.immodules", NULL);
+        }
     }
 
   return result;
@@ -522,7 +535,21 @@
   else
     {
       const gchar *home;
-      str = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtkrc", NULL);
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+# if defined (__i386__)
+      str = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtkrc.32", NULL);
+# else
+      str = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtkrc.64", NULL);
+# endif
+      /* Prefer compat gtkrc if it's usable. */
+      if (!g_file_test(str, G_FILE_TEST_EXISTS))
+        {
+          g_free (str), /* continue in next line */
+#endif
+          str = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtkrc", NULL);
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+        }
+#endif
 
       gtk_rc_add_default_file (str);
       g_free (str);
@@ -530,7 +557,21 @@
       home = g_get_home_dir ();
       if (home)
 	{
-	  str = g_build_filename (home, ".gtkrc-2.0", NULL);
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+# if defined (__i386__)
+          str = g_build_filename (home, ".gtkrc-2.0.32", NULL);
+# else
+          str = g_build_filename (home, ".gtkrc-2.0.64", NULL);
+# endif
+          /* Prefer compat .gtkrc-2.0 if it's usable. */
+          if (!g_file_test(str, G_FILE_TEST_EXISTS))
+            {
+              g_free (str), /* continue in next line */
+#endif
+              str = g_build_filename (home, ".gtkrc-2.0", NULL);
+#if defined(__linux__) && ( defined(__i386__) || defined (__x86_64__) || defined(__ia64__) )
+            }
+#endif
 	  gtk_rc_add_default_file (str);
 	  g_free (str);
 	}
--- gdk-pixbuf/gdk-pixbuf-io.c~	2008-04-17 09:56:14.355807617 +0200
+++ gdk-pixbuf/gdk-pixbuf-io.c	2008-04-17 10:05:00.843749567 +0200
