Index: gtk+2.0-2.14.0/gtk/gtktreeview.c
===================================================================
--- gtk+2.0-2.14.0.orig/gtk/gtktreeview.c	2008-09-04 06:35:27.000000000 +0200
+++ gtk+2.0-2.14.0/gtk/gtktreeview.c	2008-09-05 00:36:36.000000000 +0200
@@ -9576,7 +9576,8 @@
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
