Index: gtk+2.0-2.14.0/gtk/updateiconcache.c
===================================================================
--- gtk+2.0-2.14.0.orig/gtk/updateiconcache.c	2008-09-04 06:35:27.000000000 +0200
+++ gtk+2.0-2.14.0/gtk/updateiconcache.c	2008-09-05 00:36:37.000000000 +0200
@@ -660,7 +660,7 @@
 		      directories = g_list_append (directories, g_strdup (subdir));
 		    }
 		  else
-		    dir_index = 0xffff;
+		    continue;
 		}
 		
 	      image = g_new0 (Image, 1);
