--- gtk+2.0-2.13.6.orig/modules/printbackends/cups/gtkprintbackendcups.c
+++ gtk+2.0-2.13.6/modules/printbackends/cups/gtkprintbackendcups.c
@@ -319,8 +319,9 @@
  
   /* TODO: check if it is a ps or pdf printer */
   
-  surface = cairo_ps_surface_create_for_stream  (_cairo_write_to_cups, cache_io, width, height);
+  surface = cairo_pdf_surface_create_for_stream  (_cairo_write_to_cups, cache_io, width, height);
 
+  /*
   ppd_file = gtk_printer_cups_get_ppd (GTK_PRINTER_CUPS (printer));
 
   if (ppd_file != NULL)
@@ -336,6 +337,7 @@
 
   if (level == 3)
     cairo_ps_surface_restrict_to_level (surface, CAIRO_PS_LEVEL_3);
+  */
 
   /* TODO: DPI from settings object? */
   cairo_surface_set_fallback_resolution (surface, 300, 300);
--- gtk+2.0-2.13.6.orig/modules/printbackends/lpr/gtkprintbackendlpr.c
+++ gtk+2.0-2.13.6/modules/printbackends/lpr/gtkprintbackendlpr.c
@@ -209,7 +209,7 @@
 {
   cairo_surface_t *surface;
   
-  surface = cairo_ps_surface_create_for_stream (_cairo_write, cache_io, width, height);
+  surface = cairo_pdf_surface_create_for_stream (_cairo_write, cache_io, width, height);
 
   /* TODO: DPI from settings object? */
   cairo_surface_set_fallback_resolution (surface, 300, 300);
