Index: gtk+2.0-2.15.2/gdk/directfb/gdktestutils-directfb.c
===================================================================
--- gtk+2.0-2.15.2.orig/gdk/directfb/gdktestutils-directfb.c	2009-01-27 05:03:31.000000000 +0100
+++ gtk+2.0-2.15.2/gdk/directfb/gdktestutils-directfb.c	2009-01-27 11:21:38.000000000 +0100
@@ -175,7 +175,6 @@
  
   evt.clazz      = DFEC_WINDOW;
   evt.type       = (key_pressrelease == GDK_KEY_PRESS) ? DWET_KEYDOWN : DWET_KEYUP;
-  evt.flags      = DWEF_NONE;
   evt.window_id  = impl->dfb_id;
   evt.x          = MAX(x, 0);
   evt.y          = MAX(y, 0);
@@ -241,7 +240,6 @@
 
   evt.clazz      = DFEC_WINDOW;
   evt.type       = (button_pressrelease == GDK_BUTTON_PRESS) ? DWET_BUTTONDOWN : DWET_BUTTONUP;
-  evt.flags      = DWEF_NONE;
   evt.window_id  = impl->dfb_id;
   evt.x          = MAX(x, 0);
   evt.y          = MAX(y, 0); 
Index: gtk+2.0-2.15.2/gdk/directfb/gdkkeys-directfb.c
===================================================================
--- gtk+2.0-2.15.2.orig/gdk/directfb/gdkkeys-directfb.c	2009-01-27 11:21:37.000000000 +0100
+++ gtk+2.0-2.15.2/gdk/directfb/gdkkeys-directfb.c	2009-01-27 11:21:38.000000000 +0100
@@ -1812,6 +1812,12 @@
 	g_array_append_val (array, key);
 }
 
+gboolean
+gdk_keymap_get_caps_lock_state (GdkKeymap *keymap)
+{
+    return FALSE;
+}
+
 /**
  * gdk_keymap_get_entries_for_keyval:
  * @keymap: a #GdkKeymap, or %NULL to use the default keymap
Index: gtk+2.0-2.15.2/gdk/gdkaliasdef.c
===================================================================
--- gtk+2.0-2.15.2.orig/gdk/gdkaliasdef.c	2009-01-27 05:22:38.000000000 +0100
+++ gtk+2.0-2.15.2/gdk/gdkaliasdef.c	2009-01-27 11:21:38.000000000 +0100
@@ -1794,6 +1794,9 @@
 #undef gdk_keymap_get_for_display 
 extern __typeof (gdk_keymap_get_for_display) gdk_keymap_get_for_display __attribute((alias("IA__gdk_keymap_get_for_display"), visibility("default")));
 
+#undef gdk_keymap_get_caps_lock_state 
+extern __typeof (gdk_keymap_get_caps_lock_state) gdk_keymap_get_caps_lock_state __attribute((alias("IA__gdk_keymap_get_caps_lock_state"), visibility("default")));
+
 #undef gdk_keymap_lookup_key 
 extern __typeof (gdk_keymap_lookup_key) gdk_keymap_lookup_key __attribute((alias("IA__gdk_keymap_lookup_key"), visibility("default")));
 
