Index: gtk+2.0-2.14.1/gdk-pixbuf/gdk-pixbuf-io.c
===================================================================
--- gtk+2.0-2.14.1.orig/gdk-pixbuf/gdk-pixbuf-io.c	2008-09-05 10:41:34.000000000 +0200
+++ gtk+2.0-2.14.1/gdk-pixbuf/gdk-pixbuf-io.c	2008-09-05 10:41:47.000000000 +0200
@@ -877,7 +877,7 @@
 	g_set_error (error,
 		     GDK_PIXBUF_ERROR,
 		     GDK_PIXBUF_ERROR_UNKNOWN_TYPE,
-		     _("Image type '%s' is not supported",
+		     _("Image type '%s' is not supported"),
 		     image_module->module_name);
 	return FALSE;
 #endif  /* !USE_GMODULE */
