=== modified file 'gtk+-2.13.5/modules/printbackends/cups/gtkprintbackendcups.c'
--- gtk+-2.13.5/modules/printbackends/cups/gtkprintbackendcups.c	2008-07-23 16:52:29 +0000
+++ gtk+-2.13.5/modules/printbackends/cups/gtkprintbackendcups.c	2008-07-23 17:07:24 +0000
@@ -2673,6 +2673,7 @@
       char *cover_display_default[] = {N_("None"), N_("Classified"), N_("Confidential"), N_("Secret"), N_("Standard"), N_("Top Secret"), N_("Unclassified"),};
       char **cover = NULL;
       char **cover_display = NULL;
+      char **cover_display_translated = NULL;
       gint num_of_covers = 0;
       gpointer value;
       gint j;
@@ -2682,6 +2683,8 @@
       cover[num_of_covers] = NULL;
       cover_display = g_new (char *, num_of_covers + 1);
       cover_display[num_of_covers] = NULL;
+      cover_display_translated = g_new (char *, num_of_covers + 1);
+      cover_display_translated[num_of_covers] = NULL;
 
       for (i = 0; i < num_of_covers; i++)
         {
@@ -2697,11 +2700,11 @@
         }
 
       for (i = 0; i < num_of_covers; i++)
-        cover_display[i] = _(cover_display[i]);
+        cover_display_translated[i] = _(cover_display[i]);
   
       option = gtk_printer_option_new ("gtk-cover-before", "Before", GTK_PRINTER_OPTION_TYPE_PICKONE);
       gtk_printer_option_choices_from_array (option, num_of_covers,
-					 cover, cover_display);
+					 cover, cover_display_translated);
 
       if (backend->default_cover_before != NULL)
         gtk_printer_option_set (option, backend->default_cover_before);
@@ -2713,7 +2716,7 @@
 
       option = gtk_printer_option_new ("gtk-cover-after", "After", GTK_PRINTER_OPTION_TYPE_PICKONE);
       gtk_printer_option_choices_from_array (option, num_of_covers,
-					 cover, cover_display);
+					 cover, cover_display_translated);
       if (backend->default_cover_after != NULL)
         gtk_printer_option_set (option, backend->default_cover_after);
       else
@@ -2724,6 +2727,7 @@
 
       g_strfreev (cover);
       g_strfreev (cover_display);
+      g_free (cover_display_translated);
     }
 
   option = gtk_printer_option_new ("gtk-print-time", "Print at", GTK_PRINTER_OPTION_TYPE_PICKONE);


