diff -Nur gtk+-2.10.11.orig/gtk/gtkicontheme.c gtk+-2.10.11/gtk/gtkicontheme.c
--- gtk+-2.10.11.orig/gtk/gtkicontheme.c	2007-04-05 13:52:19.000000000 +0200
+++ gtk+-2.10.11/gtk/gtkicontheme.c	2007-04-05 13:52:47.000000000 +0200
@@ -2241,7 +2241,7 @@
       base_name = strip_suffix (name);
 
       hash_suffix = GPOINTER_TO_INT (g_hash_table_lookup (dir->icons, base_name));
-      g_hash_table_insert (icon_theme->all_icons, base_name, NULL);
+      g_hash_table_replace (icon_theme->all_icons, base_name, NULL);
       g_hash_table_replace (dir->icons, base_name, GUINT_TO_POINTER (hash_suffix| suffix));
     }
   
