--- trunk/gtk/gtkbutton.c	2007/07/03 16:58:42	18361
+++ trunk/gtk/gtkbutton.c	2007/07/03 17:09:42	18362
@@ -733,14 +733,14 @@
       image = g_object_ref (priv->image);
       if (image->parent)
 	gtk_container_remove (GTK_CONTAINER (image->parent), image);
-
-      priv->image = NULL;
     }
   
+  priv->image = NULL;
+
   if (GTK_BIN (button)->child)
     gtk_container_remove (GTK_CONTAINER (button),
 			  GTK_BIN (button)->child);
-  
+
   if (button->use_stock &&
       button->label_text &&
       gtk_stock_lookup (button->label_text, &item))
