--- gtk-2.12.9/gtk/gtkfilesystem.c	2008/04/08 13:58:45	19982
+++ gtk-2.12.9/gtk/gtkfilesystem.c	2008/04/08 13:59:47	19983
@@ -306,7 +306,7 @@
       pixbuf = gtk_widget_render_icon (widget,
                                        gtk_file_info_get_is_folder (info)
                                         ? GTK_STOCK_DIRECTORY : GTK_STOCK_FILE,
-                                       GTK_ICON_SIZE_SMALL_TOOLBAR,
+                                       GTK_ICON_SIZE_MENU,
                                        NULL);
       if (!pixbuf && error)
         g_set_error (error,
@@ -670,7 +670,7 @@
   if (!pixbuf)
     pixbuf = gtk_widget_render_icon (widget,
                                      GTK_STOCK_HARDDISK,
-                                     GTK_ICON_SIZE_SMALL_TOOLBAR,
+                                     GTK_ICON_SIZE_MENU,
                                      NULL);
 
   return pixbuf;
