--- trunk/gtk/gtkicontheme.c	2007/01/15 02:56:33	17153
+++ trunk/gtk/gtkicontheme.c	2007/02/26 14:09:21	17353
@@ -2203,8 +2203,8 @@
       base_name = strip_suffix (name);
 
       hash_suffix = GPOINTER_TO_INT (g_hash_table_lookup (dir->icons, base_name));
-      g_hash_table_replace (dir->icons, base_name, GUINT_TO_POINTER (hash_suffix| suffix));
       g_hash_table_insert (icon_theme->all_icons, base_name, NULL);
+      g_hash_table_replace (dir->icons, base_name, GUINT_TO_POINTER (hash_suffix| suffix));
     }
   
   g_dir_close (gdir);
