--- gtk+2.0-2.17.3/gdk/directfb/gdkdrawable-directfb.c	2009-06-04 21:18:03.000000000 +0200
+++ gtk+2.0-2.17.3/gdk/directfb/gdkdrawable-directfb.c.ubuntu	2009-07-10 10:42:15.000000000 +0200
@@ -1642,3 +1642,41 @@
 
 #define __GDK_DRAWABLE_X11_C__
 #include "gdkaliasdef.c"
+
+void _gdk_windowing_set_cairo_surface_size (cairo_surface_t *surface,
+					    int width,
+					    int height) {}
+
+void _gdk_windowing_before_process_all_updates     (void) {}
+
+void _gdk_windowing_after_process_all_updates      (void) {}
+
+void _gdk_windowing_window_process_updates_recurse (GdkWindow *window,
+                                                    GdkRegion *expose_region) {}
+
+cairo_surface_t * _gdk_windowing_create_cairo_surface (GdkDrawable *drawable,
+						       int width,
+						       int height) {}
+
+GdkRegion *_gdk_windowing_window_get_input_shape(GdkWindow  *window) {}
+
+GdkRegion *_gdk_windowing_get_shape_for_mask    (GdkBitmap *mask) {}
+
+GdkRegion *_gdk_windowing_window_get_shape      (GdkWindow  *window) {}
+void       _gdk_window_impl_new          (GdkWindow      *window,
+					  GdkWindow      *real_parent,
+					  GdkScreen      *screen,
+					  GdkVisual      *visual,
+					  GdkEventMask    event_mask,
+                                          GdkWindowAttr  *attributes,
+                                          gint            attributes_mask) {}
+
+gulong   _gdk_windowing_window_get_next_serial  (GdkDisplay *display) {}
+GdkGrabStatus _gdk_windowing_pointer_grab    (GdkWindow        *window,
+					      GdkWindow        *native,
+					      gboolean          owner_events,
+					      GdkEventMask      event_mask,
+					      GdkWindow        *confine_to,
+					      GdkCursor        *cursor,
+					      guint32           time) {}
+
