Index: gtk+2.0-2.14.0/gtk/gtkclipboard.c
===================================================================
--- gtk+2.0-2.14.0.orig/gtk/gtkclipboard.c	2008-09-04 06:35:28.000000000 +0200
+++ gtk+2.0-2.14.0/gtk/gtkclipboard.c	2008-09-05 00:36:46.000000000 +0200
@@ -299,6 +299,7 @@
 gtk_clipboard_get_for_display (GdkDisplay *display, 
 			       GdkAtom     selection)
 {
+  g_return_val_if_fail (display != NULL, NULL);
   g_return_val_if_fail (GDK_IS_DISPLAY (display), NULL);
   g_return_val_if_fail (!display->closed, NULL);
 
