--- gtk+2.0-2.17.5/gdk/directfb/gdkdrawable-directfb.c	2009-07-15 05:09:36.000000000 +0200
+++ gtk+2.0-2.17.5.ubuntu/gdk/directfb/gdkdrawable-directfb.c	2009-07-20 17:04:58.000000000 +0200
@@ -1666,5 +1666,12 @@
   return impl->cairo_surface;
 }
 
+void       _gdk_window_impl_new          (GdkWindow      *window,
+					  GdkWindow      *real_parent,
+					  GdkScreen      *screen,
+					  GdkVisual      *visual,
+					  GdkEventMask    event_mask,
+                                          GdkWindowAttr  *attributes,
+                                          gint            attributes_mask) {}
 #define __GDK_DRAWABLE_X11_C__
 #include "gdkaliasdef.c"
