Index: gtk+2.0-2.15.4/gdk/x11/gdkscreen-x11.c
===================================================================
--- gtk+2.0-2.15.4.orig/gdk/x11/gdkscreen-x11.c	2009-02-19 10:23:46.000000000 +0100
+++ gtk+2.0-2.15.4/gdk/x11/gdkscreen-x11.c	2009-02-19 10:23:52.000000000 +0100
@@ -866,9 +866,6 @@
   if (init_fake_xinerama (screen))
     return;
 
-  if (init_randr13 (screen))
-    return;
-
   if (XQueryExtension (GDK_SCREEN_XDISPLAY (screen), "XINERAMA",
 		       &opcode, &firstevent, &firsterror))
     {
